filebeat/registrar/registrar.go

```diff
@@ -179,10 +179,9 @@ func fixStates(states []file.State) []file.State {
 		state := &states[i]
 		fixState(state)
 
-		id := state.ID()
-		old, exists := idx[id]
+		old, exists := idx[state.Id]
 		if !exists {
-			idx[id] = state
+			idx[state.Id] = state
```

Comment:
If the `IdentifierName` is changed we also change the ID. Do we need to compensate for this in the registrar? E.g. by having a map of maps like `map[identifier]map[id]state` ?

What could happen if we don't adjust for it? Maybe we end up with duplicate entries for files because the ID was changed after the identifier type was changed?

Reply:
I am removing the outdated state in the `log` input: https://github.com/elastic/beats/pull/18748/files#diff-8f0e85354c10fc4b55e2f144249abc02R185 Do you think we need more cleaning up besides this?

Reply:
I wonder if the removal can lead to problems. By removing we rely on another state update event. Otherwise the entry is gone for good and a filebeat restart will force recollection.

Even with state update event, the remove+reinsert operation is not atmoic. The state update event can even be blocked for an undetermined amount of time (or even get lost if the input is stopped via autodiscovery), which leaves us with a file without state in the registry. => recollection